File: src/main/java/org/styl/gravitus/ui/ToolBar.java
package org.styl.gravitus.ui;

import java.awt.event.ActionListener;

import javax.swing.ButtonGroup;
import javax.swing.ImageIcon;
import javax.swing.JCheckBoxMenuItem;
import javax.swing.JMenu;
import javax.swing.JMenuBar;
import javax.swing.JMenuItem;
import javax.swing.JRadioButtonMenuItem;

import org.styl.gravitus.Specs;
import org.styl.gravitus.SystemLocations;

import lombok.Getter;
import lombok.Setter;

@SuppressWarnings("serial")
@Getter
@Setter
public class ToolBar extends JMenuBar {

	private ActionListener listener;

	private JMenuItem start;
	private JMenuItem pause;
	private JMenuItem stop;
	private JMenu fpsMenu;
	private JMenu prefsMenu;

	public ToolBar() {
		super();
	}

	public void init() {

		// File
		JMenu fileMenu = new JMenu("File");

		JMenuItem exit = new JMenuItem("Exit");
		exit.setIcon(new ImageIcon(SystemLocations.ICONS_FOLDER + "exit.png"));
		exit.setActionCommand("exit");
		exit.addActionListener(listener);
		fileMenu.add(exit);

		add(fileMenu);

		// Simulation
		JMenu simMenu = new JMenu("Simulation");

		start = new JMenuItem("Start");
		start.setIcon(new ImageIcon(SystemLocations.ICONS_FOLDER + "start.png"));
		start.setActionCommand("start");
		start.addActionListener(listener);
		simMenu.add(start);

		pause = new JMenuItem("Pause");
		pause.setIcon(new ImageIcon(SystemLocations.ICONS_FOLDER + "pause.png"));
		pause.setEnabled(false);
		pause.setActionCommand("pause");
		pause.addActionListener(listener);
		simMenu.add(pause);

		stop = new JMenuItem("Stop");
		stop.setIcon(new ImageIcon(SystemLocations.ICONS_FOLDER + "stop.png"));
		stop.setEnabled(false);
		stop.setActionCommand("stop");
		stop.addActionListener(listener);
		simMenu.add(stop);

		add(simMenu);

		// Preferences
		prefsMenu = new JMenu("Preferences");
		prefsMenu.setEnabled(false);

		JCheckBoxMenuItem trails = new JCheckBoxMenuItem("Track Orbits");
		trails.setSelected(Specs.instance.orbitTrails);
		trails.setActionCommand("trails");
		trails.addActionListener(listener);
		prefsMenu.add(trails);

		JCheckBoxMenuItem orbitPath = new JCheckBoxMenuItem("Fixed Orbits");
		orbitPath.setSelected(Specs.instance.orbitsFixed);
		orbitPath.setActionCommand("orbit_path");
		orbitPath.addActionListener(listener);
		prefsMenu.add(orbitPath);

		fpsMenu = new JMenu("FPS");

		ButtonGroup fpsGroup = new ButtonGroup();

		JRadioButtonMenuItem fps15 = new JRadioButtonMenuItem("15");
		fps15.setActionCommand("change_fps");
		fps15.addActionListener(listener);
		fpsMenu.add(fps15);
		fpsGroup.add(fps15);

		JRadioButtonMenuItem fps20 = new JRadioButtonMenuItem("20");
		fps20.setActionCommand("change_fps");
		fps20.addActionListener(listener);
		fpsMenu.add(fps20);
		fpsGroup.add(fps20);

		JRadioButtonMenuItem fps30 = new JRadioButtonMenuItem("30");
		fps30.setActionCommand("change_fps");
		fps30.addActionListener(listener);
		fpsMenu.add(fps30);
		fpsGroup.add(fps30);

		JRadioButtonMenuItem fps40 = new JRadioButtonMenuItem("40");
		fps40.setSelected(true);
		fps40.setActionCommand("change_fps");
		fps40.addActionListener(listener);
		fpsMenu.add(fps40);
		fpsGroup.add(fps40);

		JRadioButtonMenuItem fps60 = new JRadioButtonMenuItem("60");
		fps60.setActionCommand("change_fps");
		fps60.addActionListener(listener);
		fpsMenu.add(fps60);
		fpsGroup.add(fps60);

		JRadioButtonMenuItem fps120 = new JRadioButtonMenuItem("120");
		fps120.setActionCommand("change_fps");
		fps120.addActionListener(listener);
		fpsMenu.add(fps120);
		fpsGroup.add(fps120);

		JRadioButtonMenuItem fps240 = new JRadioButtonMenuItem("240");
		fps240.setActionCommand("change_fps");
		fps240.addActionListener(listener);
		fpsMenu.add(fps240);
		fpsGroup.add(fps240);

		prefsMenu.add(fpsMenu);

		add(prefsMenu);
	}
}

File: src/main/java/org/styl/gravitus/ui/ToolBar.kt
package org.styl.gravitus.ui

import org.styl.gravitus.Specs
import org.styl.gravitus.SystemLocations
import java.awt.event.ActionListener
import javax.swing.*

@Suppress("serial")
class ToolBar : JMenuBar() {

	var listener: ActionListener? = null

	lateinit var start: JMenuItem
		private set
	lateinit var pause: JMenuItem
		private set
	lateinit var stop: JMenuItem
		private set
	lateinit var fpsMenu: JMenu
		private set
	lateinit var prefsMenu: JMenu
		private set

	val stageMenu = JMenu("Stages")

	fun init(stageNames: List<String>) {
		// File menu
		val fileMenu = JMenu("File")
		val exit = JMenuItem("Exit").apply {
			icon = ImageIcon(SystemLocations.iconsFolder + "exit.png")
			actionCommand = "exit"
			listener?.let { addActionListener(it) }
		}
		fileMenu.add(exit)
		add(fileMenu)

		// Simulation menu
		val simMenu = JMenu("Simulation")

		start = JMenuItem("Start").apply {
			icon = ImageIcon(SystemLocations.iconsFolder + "start.png")
			actionCommand = "start"
			listener?.let { addActionListener(it) }
		}
		simMenu.add(start)

		pause = JMenuItem("Pause").apply {
			icon = ImageIcon(SystemLocations.iconsFolder + "pause.png")
			isEnabled = false
			actionCommand = "pause"
			listener?.let { addActionListener(it) }
		}
		simMenu.add(pause)

		stop = JMenuItem("Stop").apply {
			icon = ImageIcon(SystemLocations.iconsFolder + "stop.png")
			isEnabled = false
			actionCommand = "stop"
			listener?.let { addActionListener(it) }
		}
		simMenu.add(stop)

		add(simMenu)

		// Stage Menu
		val stageMenu = JMenu("Stages")
		stageNames.forEach { stageName ->
			val item = JMenuItem(stageName.substringBefore(".")).apply {
				actionCommand = "select_stage"
				putClientProperty("stage_name", stageName)
				listener?.let { addActionListener(it) }
			}
			stageMenu.add(item)
		}
		add(stageMenu)

		// Preferences menu
		prefsMenu = JMenu("Preferences").apply { isEnabled = false }

		val trails = JCheckBoxMenuItem("Track Orbits").apply {
			isSelected = Specs.instance?.orbitTrails == true
			actionCommand = "trails"
			listener?.let { addActionListener(it) }
		}
		prefsMenu.add(trails)

		val orbitPath = JCheckBoxMenuItem("Fixed Orbits").apply {
			isSelected = Specs.instance?.orbitsFixed == true
			actionCommand = "orbit_path"
			listener?.let { addActionListener(it) }
		}
		prefsMenu.add(orbitPath)

		fpsMenu = JMenu("FPS")
		val fpsGroup = ButtonGroup()

		listOf(15, 20, 30, 40, 60, 120, 240).forEach { fpsValue ->
			val item = JRadioButtonMenuItem(fpsValue.toString()).apply {
				actionCommand = "change_fps"
				listener?.let { addActionListener(it) }
				if (fpsValue == 40) isSelected = true
			}
			fpsMenu.add(item)
			fpsGroup.add(item)
		}

		prefsMenu.add(fpsMenu)
		add(prefsMenu)
	}
}
